File: app/src/main/java/com/zack/enderplan/domain/fragment/AllTypesFragment.java
package com.zack.enderplan.domain.fragment;

import android.app.ActivityOptions;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.zack.enderplan.R;
import com.zack.enderplan.domain.activity.TypeDetailActivity;
import com.zack.enderplan.domain.view.AllTypesView;
import com.zack.enderplan.interactor.adapter.TypeAdapter;
import com.zack.enderplan.interactor.callback.TypeItemTouchCallback;
import com.zack.enderplan.interactor.presenter.AllTypesPresenter;

import butterknife.BindView;
import butterknife.ButterKnife;

public class AllTypesFragment extends Fragment implements AllTypesView {

    @BindView(R.id.list_all_types)
    RecyclerView mAllTypesList;

    private AllTypesPresenter mAllTypesPresenter;

    public AllTypesFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mAllTypesPresenter = new AllTypesPresenter(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_all_types, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ButterKnife.bind(this, view);

        mAllTypesPresenter.setInitialView();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mAllTypesPresenter.detachView();
    }

    @Override
    public void showInitialView(TypeAdapter typeAdapter) {

        typeAdapter.setOnTypeItemClickListener(new TypeAdapter.OnTypeItemClickListener() {
            @Override
            public void onTypeItemClick(int position, View typeItem) {
                mAllTypesPresenter.notifyTypeItemClicked(position, typeItem);
            }
        });

        mAllTypesList.setLayoutManager(new LinearLayoutManager(getActivity()));
        mAllTypesList.setHasFixedSize(true);
        mAllTypesList.setAdapter(typeAdapter);

        TypeItemTouchCallback typeItemTouchCallback = new TypeItemTouchCallback();
        typeItemTouchCallback.setOnItemMovedListener(new TypeItemTouchCallback.OnItemMovedListener() {
            @Override
            public void onItemMoved(int fromPosition, int toPosition) {
                mAllTypesPresenter.notifyTypeSequenceChanged(fromPosition, toPosition);
            }
        });
        new ItemTouchHelper(typeItemTouchCallback).attachToRecyclerView(mAllTypesList);
    }

    @Override
    public void onTypeItemClicked(int position, View typeItem) {
        TypeDetailActivity.start(
                getActivity(),
                position,
                typeItem.findViewById(R.id.ic_type_mark),
                getResources().getString(R.string.name_type_mark_shared_element_transition)
        );
    }
}

File: app/src/main/java/net/zackzhang/app/end/view/fragment/AllTypesFragment.kt
package net.zackzhang.app.end.view.fragment

import android.os.Bundle
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.fragment_all_types.*
import net.zackzhang.app.end.App
import net.zackzhang.app.end.R
import net.zackzhang.app.end.injector.component.DaggerAllTypesComponent
import net.zackzhang.app.end.injector.module.AllTypesPresenterModule
import net.zackzhang.app.end.presenter.AllTypesPresenter
import net.zackzhang.app.end.view.activity.TypeDetailActivity
import net.zackzhang.app.end.view.adapter.TypeListAdapter
import net.zackzhang.app.end.view.contract.AllTypesViewContract
import javax.inject.Inject

class AllTypesFragment : BaseListFragment(), AllTypesViewContract {

    @Inject
    lateinit var mAllTypesPresenter: AllTypesPresenter

    override fun onInjectPresenter() {
        DaggerAllTypesComponent.builder()
                .allTypesPresenterModule(AllTypesPresenterModule(this))
                .appComponent(App.appComponent)
                .build()
                .inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) =
            inflater.inflate(R.layout.fragment_all_types, container, false)!!

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mAllTypesPresenter.attach()
    }

    override fun onDetach() {
        super.onDetach()
        mAllTypesPresenter.detach()
    }

    override fun showInitialView(typeListAdapter: TypeListAdapter, itemTouchHelper: ItemTouchHelper) {
        list_type.setHasFixedSize(true)
        list_type.adapter = typeListAdapter
        list_type.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                onListScrolled(dy)
                mAllTypesPresenter.notifyPlanListScrolled(
                        !list_type.canScrollVertically(-1),
                        !list_type.canScrollVertically(1)
                )
            }
        })
        itemTouchHelper.attachToRecyclerView(list_type)
    }

    override fun onTypeItemClicked(position: Int, typeItem: View) {
        val typeMarkIcon = typeItem.findViewById<View>(R.id.ic_type_mark)
        TypeDetailActivity.start(
                activity!!,
                position,
                true,
                typeMarkIcon,
                typeMarkIcon.transitionName
        )
    }

    override fun onTypeCreated(scrollTo: Int) {
        list_type.scrollToPosition(scrollTo)
    }
}
